xds/src/main/java/com/linecorp/armeria/xds/client/endpoint/ClusterEntry.java

```diff
@@ -20,46 +20,52 @@
 
 import java.util.List;
 import java.util.concurrent.CompletableFuture;
-import java.util.function.Consumer;
 
 import com.google.common.base.MoreObjects;
 import com.google.common.collect.ImmutableList;
 
 import com.linecorp.armeria.client.ClientRequestContext;
 import com.linecorp.armeria.client.Endpoint;
-import com.linecorp.armeria.client.endpoint.EndpointGroup;
 import com.linecorp.armeria.common.annotation.Nullable;
 import com.linecorp.armeria.common.util.AsyncCloseable;
 import com.linecorp.armeria.xds.ClusterSnapshot;
 import com.linecorp.armeria.xds.EndpointSnapshot;
 
-final class ClusterEntry implements Consumer<List<Endpoint>>, AsyncCloseable {
+import io.netty.util.concurrent.EventExecutor;
 
-    private final EndpointGroup endpointGroup;
-    private final LoadBalancer loadBalancer;
-    private List<Endpoint> endpoints = ImmutableList.of();
+final class ClusterEntry implements AsyncCloseable {
 
-    ClusterEntry(ClusterSnapshot clusterSnapshot, ClusterManager clusterManager) {
-        final EndpointSnapshot endpointSnapshot = clusterSnapshot.endpointSnapshot();
-        assert endpointSnapshot != null;
-        loadBalancer = new SubsetLoadBalancer(clusterSnapshot);
+    private final EndpointsPool endpointsPool;
+    private final LoadBalancer loadBalancer = new SubsetLoadBalancer();
+    private final ClusterManager clusterManager;
+    private final EventExecutor eventExecutor;
+    private List<Endpoint> endpoints = ImmutableList.of();
 
-        // The order of adding listeners is important
-        endpointGroup = XdsEndpointUtil.convertEndpointGroup(clusterSnapshot);
-        endpointGroup.addListener(this, true);
-        endpointGroup.addListener(clusterManager, true);
+    ClusterEntry(ClusterSnapshot clusterSnapshot,
+                 ClusterManager clusterManager, EventExecutor eventExecutor) {
+        this.clusterManager = clusterManager;
+        this.eventExecutor = eventExecutor;
+        endpointsPool = new EndpointsPool(this, eventExecutor);
+        updateClusterSnapshot(clusterSnapshot);
     }
 
     @Nullable
     Endpoint selectNow(ClientRequestContext ctx) {
         return loadBalancer.selectNow(ctx);
     }
 
-    @Override
-    public void accept(List<Endpoint> endpoints) {
-        this.endpoints = ImmutableList.copyOf(endpoints);
-        final PrioritySet prioritySet = new PrioritySet(endpoints);
+    void updateClusterSnapshot(ClusterSnapshot clusterSnapshot) {
+        final EndpointSnapshot endpointSnapshot = clusterSnapshot.endpointSnapshot();
+        assert endpointSnapshot != null;
+        endpointsPool.updateClusterSnapshot(clusterSnapshot);
```

Comment:
Optional) The cross-coupling between `EndpointsPool` and `ClusterEntry` could be eliminated by passing a callback to `endpointsPool.updateClusterSnapshot()`.
```suggestion
        endpointsPool.updateClusterSnapshot(clusterSnapshot, endpoints -> {
            accept(clusterSnapshot, endpoints);
        });
```

Reply:
Good point, done 👍